### DISCUSSION TITLE
Is it possible to watch realtime subscribers of a certain subject?

### DISCUSSION THREAD

**Discussion by tpoxa** [nats-io/nats.go#1051] [Q&A] [answered]
I need to build something like a scheduler which watches for subscribers of a list of subjects and spawns a new instance of s subscriber automatically.

Is this possible using JetStream?

Thanks.

**Comment by derekcollison** [comment, marked as answer]
JetStream can do WorkQueue semantics where a job or request is submitted and you can have a horizontally scalable set of workers process them. Would that work or am I misunderstanding the scenario?

**Reply by tpoxa** [reply]
Thanks for the reply.

My workers use a set of known channels that they are listening for commands from a manager.
Workers work only with NATS. At least that was a design...

The manager may ask a worker to spawn a new subprocess by consuming new subjects generated on the fly. 

The problem is when workers suddenly restart, they don't know what they have to subscribe to again, except the same list of common known lists of subjects. The Manager only knows about spawned subjects, but the manager does not know if it needs to ask some available worker to respawn it again because I can not find the way how a manager can tell if the subject lost all consumers...

Maybe an example will help.

A worker during bootstrap subscribes to predefined subjects.


- **components.create** - receiving a message here means the Manager asks to spawn a new worker with a payload given. Workers start listening to a new subject in a goroutine 
  - "component.id8206582334.*" (special a long-running task is now running here and listens for its own subcommands)
   
 
- **components.discovery** - reply to the manager about available worker and all its spawned routines
 


The problem is when my worker restarts, it knows only about **components.create** and **components.discovery** subjects to listen. An know nothing about **component.id8206582334.***
The manager needs to ask for some available workers to deploy   - "component.id8206582334.*"  again.

---
It looks like I need some service discovery service like consul or ETCD with their distributed locks...


Sorry for the confusion.

**Reply by derekcollison** [reply]
Possibly, but we also might be overthinking the problem. 

What do the workers actually do? What role does the discovered subjects play in the context of the worker?

I think there might be a simpler approach.

**Reply by tpoxa** [reply]
Hey Derek. Thanks for not giving up :)

I build  Flow-based programming system  (https://en.wikipedia.org/wiki/Flow-based_programming) 

Each worker supports a certain type of graph node (read functionality).  So the list of workers installed may vary depending on a user's needs.  The Discovery subject is made to understand a palette of functionality available.

 When a user adds a desired node to the graph, it becomes an autonomous **instance** (component.id8206582334.*), which reads its incoming messages, processes them specifically and passes them to the next node. 

That's where I am now.

So as a possible simple solution - the manager, from time to time, publishes a ping message which should be answered by all instances in some period of time. The manager gets all of them back, matches them with persistency storage, and now it knows which instances are not functioning (worker was restarted).  The manager sends **"components.create"** message to the available workers to instantiate them. 

I just don't like that "time to time"...

**Reply by tpoxa** [reply]
As an example of a simple graph
<img width="1116" alt="Screenshot 2022-08-18 at 20 50 54" src="https://user-images.githubusercontent.com/431546/185753980-82c1edb0-8d6d-4447-91c5-f149ddf7d6d6.png">

**Reply by derekcollison** [reply]
Would it make sense that when a node is created (with a sort of UUID/NUID) that a NATS KV entry is created that represents all the subjects it should be interested in?

Then when a node restarts, it first gets its state from the KV and listens on those subjects (could be JetStream consumers with backing streams) for any missed messages?

**Reply by tpoxa** [reply]
Yeah, that's might work out. Thanks. Just need to take care that this node id won't be lost during scale up and scale down. 
Thanks, @derekcollison; I think this is the way.

**Reply by derekcollison** [reply]
Keep us posted!
